File: output/java/tests/51_hsm_persist.java
import java.util.*;


import java.util.*;

class HSMPersistFrameEvent {
    String _message;
    HashMap<String, Object> _parameters;

    HSMPersistFrameEvent(String message) {
        this._message = message;
        this._parameters = null;
    }

    HSMPersistFrameEvent(String message, HashMap<String, Object> parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}

class HSMPersistFrameContext {
    HSMPersistFrameEvent _event;
    Object _return;
    HashMap<String, Object> _data;

    HSMPersistFrameContext(HSMPersistFrameEvent event, Object defaultReturn) {
        this._event = event;
        this._return = defaultReturn;
        this._data = new HashMap<>();
    }
}

class HSMPersistCompartment {
    String state;
    HashMap<String, Object> state_args;
    HashMap<String, Object> state_vars;
    HashMap<String, Object> enter_args;
    HashMap<String, Object> exit_args;
    HSMPersistFrameEvent forward_event;
    HSMPersistCompartment parent_compartment;

    HSMPersistCompartment(String state) {
        this.state = state;
        this.state_args = new HashMap<>();
        this.state_vars = new HashMap<>();
        this.enter_args = new HashMap<>();
        this.exit_args = new HashMap<>();
        this.forward_event = null;
        this.parent_compartment = null;
    }

    HSMPersistCompartment copy() {
        HSMPersistCompartment c = new HSMPersistCompartment(this.state);
        c.state_args = new HashMap<>(this.state_args);
        c.state_vars = new HashMap<>(this.state_vars);
        c.enter_args = new HashMap<>(this.enter_args);
        c.exit_args = new HashMap<>(this.exit_args);
        c.forward_event = this.forward_event;
        c.parent_compartment = this.parent_compartment;
        return c;
    }
}

class HSMPersist {
    private ArrayList<HSMPersistCompartment> _state_stack;
    private HSMPersistCompartment __compartment;
    private HSMPersistCompartment __next_compartment;
    private ArrayList<HSMPersistFrameContext> _context_stack;
    public int parent_count_init = 100;

    public HSMPersist() {
        _state_stack = new ArrayList<>();
        _context_stack = new ArrayList<>();
        __compartment = new HSMPersistCompartment("Parent");
        __next_compartment = null;
        HSMPersistFrameEvent __frame_event = new HSMPersistFrameEvent("$>");
        HSMPersistFrameContext __ctx = new HSMPersistFrameContext(__frame_event, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    private void __kernel(HSMPersistFrameEvent __e) {
        __router(__e);
        while (__next_compartment != null) {
            var next_compartment = __next_compartment;
            __next_compartment = null;
            HSMPersistFrameEvent exit_event = new HSMPersistFrameEvent("<$");
            __router(exit_event);
            __compartment = next_compartment;
            if (__compartment.forward_event == null) {
                HSMPersistFrameEvent enter_event = new HSMPersistFrameEvent("$>");
                __router(enter_event);
            } else {
                var forward_event = __compartment.forward_event;
                __compartment.forward_event = null;
                if (forward_event._message.equals("$>")) {
                    __router(forward_event);
                } else {
                    HSMPersistFrameEvent enter_event = new HSMPersistFrameEvent("$>");
                    __router(enter_event);
                    __router(forward_event);
                }
            }
        }
    }

    private void __router(HSMPersistFrameEvent __e) {
        String state_name = __compartment.state;
        if (state_name.equals("Parent")) {
            _state_Parent(__e);
        } else if (state_name.equals("Child")) {
            _state_Child(__e);
        }
    }

    private void __transition(HSMPersistCompartment next) {
        __next_compartment = next;
    }

    public void increment_child() {
        HSMPersistFrameEvent __e = new HSMPersistFrameEvent("increment_child");
        HSMPersistFrameContext __ctx = new HSMPersistFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    public void increment_parent() {
        HSMPersistFrameEvent __e = new HSMPersistFrameEvent("increment_parent");
        HSMPersistFrameContext __ctx = new HSMPersistFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    public int get_child_count() {
        HSMPersistFrameEvent __e = new HSMPersistFrameEvent("get_child_count");
        HSMPersistFrameContext __ctx = new HSMPersistFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        var __result = (int) _context_stack.get(_context_stack.size() - 1)._return;
        _context_stack.remove(_context_stack.size() - 1);
        return __result;
    }

    public int get_parent_count() {
        HSMPersistFrameEvent __e = new HSMPersistFrameEvent("get_parent_count");
        HSMPersistFrameContext __ctx = new HSMPersistFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        var __result = (int) _context_stack.get(_context_stack.size() - 1)._return;
        _context_stack.remove(_context_stack.size() - 1);
        return __result;
    }

    public String get_state() {
        HSMPersistFrameEvent __e = new HSMPersistFrameEvent("get_state");
        HSMPersistFrameContext __ctx = new HSMPersistFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        var __result = (String) _context_stack.get(_context_stack.size() - 1)._return;
        _context_stack.remove(_context_stack.size() - 1);
        return __result;
    }

    private void _state_Parent(HSMPersistFrameEvent __e) {
        if (__e._message.equals("get_child_count")) {
            _context_stack.get(_context_stack.size() - 1)._return = -1;
            return;
        } else if (__e._message.equals("get_parent_count")) {
            _context_stack.get(_context_stack.size() - 1)._return = (int) __compartment.state_vars.get("parent_count");
            return;
        } else if (__e._message.equals("get_state")) {
            _context_stack.get(_context_stack.size() - 1)._return = "Parent";
            return;
        } else if (__e._message.equals("increment_child")) {
        } else if (__e._message.equals("increment_parent")) {
            __compartment.state_vars.put("parent_count", (int) __compartment.state_vars.get("parent_count") + 1);
        }
    }

    private void _state_Child(HSMPersistFrameEvent __e) {
        var __sv_comp = __compartment;
        while (__sv_comp != null && !__sv_comp.state.equals("Child")) { __sv_comp = __sv_comp.parent_compartment; }
        if (__e._message.equals("$>")) {
            if (!__sv_comp.state_vars.containsKey("child_count")) {
                __sv_comp.state_vars.put("child_count", 0);
            }
        } else if (__e._message.equals("get_child_count")) {
            _context_stack.get(_context_stack.size() - 1)._return = (int) __sv_comp.state_vars.get("child_count");
            return;
        } else if (__e._message.equals("get_state")) {
            _context_stack.get(_context_stack.size() - 1)._return = "Child";
            return;
        } else if (__e._message.equals("increment_child")) {
            __sv_comp.state_vars.put("child_count", (int) __sv_comp.state_vars.get("child_count") + 1);
        }
    }

    private org.json.JSONObject __serComp(HSMPersistCompartment comp) {
        if (comp == null) return null;
        var j = new org.json.JSONObject();
        j.put("state", comp.state);
        var sv = new org.json.JSONObject();
        for (var e : comp.state_vars.entrySet()) { sv.put(e.getKey(), e.getValue()); }
        j.put("state_vars", sv);
        j.put("parent", __serComp(comp.parent_compartment));
        return j;
    }

    private static HSMPersistCompartment __deserComp(Object obj) {
        if (obj == null || obj.equals(org.json.JSONObject.NULL)) return null;
        var d = (org.json.JSONObject) obj;
        var c = new HSMPersistCompartment(d.getString("state"));
        if (d.has("state_vars")) {
            var sv = d.getJSONObject("state_vars");
            for (var k : sv.keySet()) { c.state_vars.put(k, sv.get(k)); }
        }
        if (d.has("parent") && !d.isNull("parent")) {
            c.parent_compartment = __deserComp(d.get("parent"));
        }
        return c;
    }

    public String save_state() {
        var __j = new org.json.JSONObject();
        __j.put("_compartment", __serComp(__compartment));
        var __stack = new org.json.JSONArray();
        for (var c : _state_stack) { __stack.put(__serComp(c)); }
        __j.put("_state_stack", __stack);
        __j.put("parent_count_init", parent_count_init);
        return __j.toString();
    }

    public static HSMPersist restore_state(String json) {
        var __j = new org.json.JSONObject(json);
        var __instance = new HSMPersist();
        __instance.__compartment = __deserComp(__j.get("_compartment"));
        if (__j.has("_state_stack")) {
            var __stack = __j.getJSONArray("_state_stack");
            __instance._state_stack = new ArrayList<>();
            for (int i = 0; i < __stack.length(); i++) { __instance._state_stack.add(__deserComp(__stack.get(i))); }
        }
        if (__j.has("parent_count_init")) { __instance.parent_count_init = __j.getInt("parent_count_init"); }
        return __instance;
    }
}

class Main {
    public static void main(String[] args) {
        System.out.println("=== Test 51: HSM Persistence (Java) ===");

        // Note: parent_count is a state var on $Parent, init from domain
        // For simplicity, test with child_count state var only
        var s1 = new HSMPersist();
        s1.increment_child();
        s1.increment_child();
        s1.increment_child();

        if (s1.get_child_count() != 3) {
            System.out.println("FAIL: child_count " + s1.get_child_count()); System.exit(1);
        }
        System.out.println("1. child_count: " + s1.get_child_count());

        String json = s1.save_state();
        System.out.println("2. Saved: " + json.length() + " chars");

        var s2 = HSMPersist.restore_state(json);
        if (!s2.get_state().equals("Child")) {
            System.out.println("FAIL: state " + s2.get_state()); System.exit(1);
        }
        System.out.println("3. Restored state: " + s2.get_state());

        if (s2.get_child_count() != 3) {
            System.out.println("FAIL: restored child_count " + s2.get_child_count()); System.exit(1);
        }
        System.out.println("4. child_count preserved: " + s2.get_child_count());

        s2.increment_child();
        if (s2.get_child_count() != 4) {
            System.out.println("FAIL: post-restore " + s2.get_child_count()); System.exit(1);
        }
        System.out.println("5. After increment: " + s2.get_child_count());

        System.out.println("PASS: HSM persistence works correctly");
    }
}

File: output/go/tests/51_hsm_persist.go

package main

import (
	"encoding/json"
	"fmt"
	"os"
)

// Test 51: HSM Persistence
//
// Tests that HSM parent compartment chain is properly saved and restored:
// - Child state vars preserved
// - Parent state vars preserved
// - Parent compartment chain intact after restore
// - Forwarding to parent still works after restore

type HSMPersistFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type HSMPersistFrameContext struct {
    _event  HSMPersistFrameEvent
    _return any
    _data   map[string]any
}

type HSMPersistCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *HSMPersistFrameEvent
    parentCompartment *HSMPersistCompartment
}

func newHSMPersistCompartment(state string) *HSMPersistCompartment {
    return &HSMPersistCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *HSMPersistCompartment) copy() *HSMPersistCompartment {
    nc := &HSMPersistCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type HSMPersist struct {
    _state_stack []*HSMPersistCompartment
    __compartment *HSMPersistCompartment
    __next_compartment *HSMPersistCompartment
    _context_stack []HSMPersistFrameContext
}

func NewHSMPersist() *HSMPersist {
    s := &HSMPersist{}
    s._state_stack = make([]*HSMPersistCompartment, 0)
    s._context_stack = make([]HSMPersistFrameContext, 0)
    // HSM: Create parent compartment chain
    __parent_comp_0 := newHSMPersistCompartment("Parent")
    __parent_comp_0.parentCompartment = nil
    __parent_comp_0.stateVars["parent_count"] = 100
    s.__compartment = newHSMPersistCompartment("Child")
    s.__compartment.parentCompartment = __parent_comp_0
    s.__next_compartment = nil
    __frame_event := HSMPersistFrameEvent{_message: "$>", _parameters: nil}
    __ctx := HSMPersistFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *HSMPersist) __kernel(__e *HSMPersistFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &HSMPersistFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &HSMPersistFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &HSMPersistFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *HSMPersist) __router(__e *HSMPersistFrameEvent) {
    switch s.__compartment.state {
    case "Child":
        s._state_Child(__e)
    case "Parent":
        s._state_Parent(__e)
    }
}

func (s *HSMPersist) __transition(next *HSMPersistCompartment) {
    s.__next_compartment = next
}

func (s *HSMPersist) IncrementChild() {
    __e := HSMPersistFrameEvent{_message: "IncrementChild"}
    __ctx := HSMPersistFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HSMPersist) IncrementParent() {
    __e := HSMPersistFrameEvent{_message: "IncrementParent"}
    __ctx := HSMPersistFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HSMPersist) GetChildCount() int {
    __e := HSMPersistFrameEvent{_message: "GetChildCount"}
    __ctx := HSMPersistFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *HSMPersist) GetParentCount() int {
    __e := HSMPersistFrameEvent{_message: "GetParentCount"}
    __ctx := HSMPersistFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *HSMPersist) GetState() string {
    __e := HSMPersistFrameEvent{_message: "GetState"}
    __ctx := HSMPersistFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *HSMPersist) _state_Parent(__e *HSMPersistFrameEvent) {
    __sv_comp := s.__compartment
    for __sv_comp != nil && __sv_comp.state != "Parent" { __sv_comp = __sv_comp.parentCompartment }
    if __e._message == "$>" {
        if _, ok := __sv_comp.stateVars["parent_count"]; !ok {
            __sv_comp.stateVars["parent_count"] = 100
        }
    } else if __e._message == "GetParentCount" {
        s._context_stack[len(s._context_stack)-1]._return = __sv_comp.stateVars["parent_count"].(int)
        return
    } else if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "Parent"
        return
    } else if __e._message == "IncrementParent" {
        __sv_comp.stateVars["parent_count"] = __sv_comp.stateVars["parent_count"].(int) + 1
    }
}

func (s *HSMPersist) _state_Child(__e *HSMPersistFrameEvent) {
    __sv_comp := s.__compartment
    for __sv_comp != nil && __sv_comp.state != "Child" { __sv_comp = __sv_comp.parentCompartment }
    if __e._message == "$>" {
        if _, ok := __sv_comp.stateVars["child_count"]; !ok {
            __sv_comp.stateVars["child_count"] = 0
        }
    } else if __e._message == "GetChildCount" {
        s._context_stack[len(s._context_stack)-1]._return = __sv_comp.stateVars["child_count"].(int)
        return
    } else if __e._message == "GetParentCount" {
        // Forward to parent
        s._state_Parent(__e)
    } else if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "Child"
        return
    } else if __e._message == "IncrementChild" {
        __sv_comp.stateVars["child_count"] = __sv_comp.stateVars["child_count"].(int) + 1
    } else if __e._message == "IncrementParent" {
        // Forward to parent
        s._state_Parent(__e)
    }
}

// --- Persist helpers with HSM parent compartment chain ---

type hsmCompSnap struct {
	State     string         `json:"state"`
	StateVars map[string]any `json:"state_vars,omitempty"`
	Parent    *hsmCompSnap   `json:"parent,omitempty"`
}

type hsmPersistSnapshot struct {
	Compartment hsmCompSnap `json:"compartment"`
}

func saveHSMComp(c *HSMPersistCompartment) hsmCompSnap {
	vars := make(map[string]any)
	for k, v := range c.stateVars {
		vars[k] = v
	}
	snap := hsmCompSnap{State: c.state, StateVars: vars}
	if c.parentCompartment != nil {
		parent := saveHSMComp(c.parentCompartment)
		snap.Parent = &parent
	}
	return snap
}

func saveHSMPersist(s *HSMPersist) string {
	snap := hsmPersistSnapshot{
		Compartment: saveHSMComp(s.__compartment),
	}
	b, err := json.Marshal(snap)
	if err != nil {
		panic(err)
	}
	return string(b)
}

func restoreHSMComp(cs hsmCompSnap) *HSMPersistCompartment {
	c := newHSMPersistCompartment(cs.State)
	for k, v := range cs.StateVars {
		if f, ok := v.(float64); ok {
			c.stateVars[k] = int(f)
		} else {
			c.stateVars[k] = v
		}
	}
	if cs.Parent != nil {
		c.parentCompartment = restoreHSMComp(*cs.Parent)
	}
	return c
}

func restoreHSMPersist(jsonStr string) *HSMPersist {
	var snap hsmPersistSnapshot
	if err := json.Unmarshal([]byte(jsonStr), &snap); err != nil {
		panic(err)
	}
	s := &HSMPersist{}
	s._state_stack = make([]*HSMPersistCompartment, 0)
	s._context_stack = make([]HSMPersistFrameContext, 0)
	s.__compartment = restoreHSMComp(snap.Compartment)
	s.__next_compartment = nil
	return s
}

func main() {
	fmt.Println("=== Test 51: HSM Persistence (Go) ===")

	// Create system and modify state vars at both levels
	s1 := NewHSMPersist()

	// Verify initial state
	if s1.GetState() != "Child" {
		fmt.Printf("FAIL: Expected Child, got %s\n", s1.GetState())
		os.Exit(1)
	}
	if s1.GetChildCount() != 0 {
		fmt.Printf("FAIL: Expected 0, got %d\n", s1.GetChildCount())
		os.Exit(1)
	}
	if s1.GetParentCount() != 100 {
		fmt.Printf("FAIL: Expected 100, got %d\n", s1.GetParentCount())
		os.Exit(1)
	}
	fmt.Println("1. Initial state verified: Child with child_count=0, parent_count=100")

	// Modify state vars at both levels
	s1.IncrementChild()
	s1.IncrementChild()
	s1.IncrementChild() // child_count = 3
	s1.IncrementParent()
	s1.IncrementParent() // parent_count = 102

	if s1.GetChildCount() != 3 {
		fmt.Printf("FAIL: Expected 3, got %d\n", s1.GetChildCount())
		os.Exit(1)
	}
	if s1.GetParentCount() != 102 {
		fmt.Printf("FAIL: Expected 102, got %d\n", s1.GetParentCount())
		os.Exit(1)
	}
	fmt.Printf("2. After increments: child_count=%d, parent_count=%d\n", s1.GetChildCount(), s1.GetParentCount())

	// Save state
	data := saveHSMPersist(s1)
	fmt.Printf("3. Saved state: %s\n", data)

	// Restore to new instance
	s2 := restoreHSMPersist(data)

	// Verify restored state
	if s2.GetState() != "Child" {
		fmt.Printf("FAIL: Expected Child after restore, got %s\n", s2.GetState())
		os.Exit(1)
	}
	fmt.Printf("4. Restored state: %s\n", s2.GetState())

	// Verify child state var preserved
	if s2.GetChildCount() != 3 {
		fmt.Printf("FAIL: Expected child_count=3, got %d\n", s2.GetChildCount())
		os.Exit(1)
	}
	fmt.Printf("5. Child state var preserved: child_count=%d\n", s2.GetChildCount())

	// Verify parent state var preserved (requires parent compartment chain)
	if s2.GetParentCount() != 102 {
		fmt.Printf("FAIL: Expected parent_count=102, got %d\n", s2.GetParentCount())
		os.Exit(1)
	}
	fmt.Printf("6. Parent state var preserved: parent_count=%d\n", s2.GetParentCount())

	// Verify state machine still works after restore
	s2.IncrementChild()
	s2.IncrementParent()
	if s2.GetChildCount() != 4 {
		fmt.Printf("FAIL: Expected 4, got %d\n", s2.GetChildCount())
		os.Exit(1)
	}
	if s2.GetParentCount() != 103 {
		fmt.Printf("FAIL: Expected 103, got %d\n", s2.GetParentCount())
		os.Exit(1)
	}
	fmt.Printf("7. After post-restore increments: child_count=%d, parent_count=%d\n", s2.GetChildCount(), s2.GetParentCount())

	fmt.Println("PASS: HSM persistence works correctly")
}
